File: 08-longest-substring-k-zeros/solution.go
// Longest Substring with at Most K Zeros
// Time: O(n), Space: O(1)

package main

import "fmt"

func longestOnes(s string, k int) int {
    maxLen := 0
    left := 0
    zeroCount := 0
    
    for right := 0; right < len(s); right++ {
        if s[right] == '0' {
            zeroCount++
        }
        
        for zeroCount > k {
            if s[left] == '0' {
                zeroCount--
            }
            left++
        }
        
        if right - left + 1 > maxLen {
            maxLen = right - left + 1
        }
    }
    
    return maxLen
}

// Test
func main() {
    fmt.Println(longestOnes("1101100111", 2))  // Output: 8
    fmt.Println(longestOnes("000111", 2))      // Output: 5
}

File: 08-longest-substring-k-zeros/solution.cs
// Longest Substring with at Most K Zeros
// Time: O(n), Space: O(1)

using System;

public class Solution {
    public int LongestOnes(string s, int k) {
        int maxLen = 0;
        int left = 0;
        int zeroCount = 0;
        
        for (int right = 0; right < s.Length; right++) {
            if (s[right] == '0') {
                zeroCount++;
            }
            
            while (zeroCount > k) {
                if (s[left] == '0') {
                    zeroCount--;
                }
                left++;
            }
            
            maxLen = Math.Max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
    
    // Test
    public static void Main(string[] args) {
        Solution sol = new Solution();
        Console.WriteLine(sol.LongestOnes("1101100111", 2)); // Output: 8
        Console.WriteLine(sol.LongestOnes("000111", 2));     // Output: 5
    }
}
